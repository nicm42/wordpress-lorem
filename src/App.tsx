import { useState, useEffect } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import Post from './components/Post';

const App = () => {
  const [posts, setPosts] = useState<any[]>([]);

  const fetchData = async () => {
    try {
      const link =
        'https://public-api.wordpress.com/rest/v1.1/sites/nictesting935058505.wordpress.com/posts/?pretty=true';
      const request = await fetch(link);
      if (request.ok) {
        const data = await request.json();
        setPosts(data.posts);
        console.log(data.posts);
      }
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  return (
    <main>
      <h1>Alternative Ipsum Posts</h1>
      {posts.map((post: any[]) => (
        <Post post={post} key={post.ID} />
      ))}
    </main>
  );
};

export default App;
