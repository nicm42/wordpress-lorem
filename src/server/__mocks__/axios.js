export default {
  //get: jest.fn().mockResolvedValue({ data: {} }),
  get: jest.fn(),
};
