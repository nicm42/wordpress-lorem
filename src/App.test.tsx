import {
  render,
  screen,
  cleanup,
  waitFor,
  fireEvent,
} from '@testing-library/react';
import axios from 'axios';
import App from './App';
import testPost from './dummyData/dummy-posts2.json';

describe('App tests', () => {
  it('renders without crashing', async () => {
    render(<App />);
    const header = await waitFor(() =>
      screen.getByText('Alternative Ipsum Posts')
    );
    expect(header).toBeInTheDocument();
  });
});

describe('Post API test', () => {
  jest.mock('axios');
  const mockedAxios = axios as jest.Mocked<typeof axios>;

  beforeEach(() => jest.resetAllMocks());
  afterEach(cleanup);

  it('gets data from the API', async () => {
    mockedAxios.get.mockResolvedValueOnce(testPost);
    render(<App />);
    //await waitFor(() => screen.debug());
    let loadingDiv = screen.queryByTestId('loading');
    expect(loadingDiv).toBeInTheDocument();
    await waitFor(() => expect(mockedAxios.get).toHaveBeenCalledTimes(1));
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'https://public-api.wordpress.com/rest/v1.1/sites/nictesting935058505.wordpress.com/posts/?pretty=true'
      )
    );
    const Title = await waitFor(() => screen.getByText('This is a test post'));
    expect(Title).toBeInTheDocument();
    loadingDiv = await waitFor(() => screen.queryByTestId('loading'));
    expect(loadingDiv).not.toBeInTheDocument();
    let Toast = await waitFor(() => screen.queryByText('Posts loaded!'));
    expect(Toast).toBeInTheDocument();
    const toastButton = await waitFor(() =>
      screen.queryByRole('button', {
        name: /Close/i,
      })
    );
    fireEvent.click(toastButton);
    Toast = await waitFor(() => screen.queryByText('Posts loaded!'));
    expect(Toast).not.toBeInTheDocument();
  });

  it('gets an error from the API', async () => {
    mockedAxios.get.mockResolvedValueOnce('error');
    render(<App />);
    await waitFor(() => expect(mockedAxios.get).toHaveBeenCalledTimes(1));
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'https://public-api.wordpress.com/rest/v1.1/sites/nictesting935058505.wordpress.com/posts/?pretty=true'
      )
    );
    const Error = await waitFor(() => screen.getByText("Couldn't fetch posts"));
    expect(Error).toBeInTheDocument();
  });
});
