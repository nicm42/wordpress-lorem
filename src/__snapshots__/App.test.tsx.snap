// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test renders the app correctly 1`] = `
<main
  className="p-3"
>
  <h1
    className="text-center pt-2 pb-2"
  >
    Alternative Ipsum Posts
  </h1>
  <div
    className="text-center"
    data-testid="loading"
  >
    <div
      className="spinner-border text-primary"
      role="status"
    >
      <span
        className="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
  <p
    className="text-danger text-center"
  >
    
  </p>
  
</main>
`;
